### PR TITLE
Jml remove nelmio

### PR DESCRIPTION
**Pull request by jmleroux**
We don't use ApiDoc and it loads too much stuff

After that, a circular dependencie appears. It was just hidden by Nelmio because it was not resolving in the same order.

And finally, after having to use a service locator, i decided to use it for all Oro Navigation listeners to avoid loading the whole world on each call. Well, if not the whole world, Twig at least should not be loaded on every REST calls.

It resulted in much lower numbers for wall time, CPU and memory usage.

And maybe a new paradigm will emerge: **For kernel listeners, we should always use a service locator**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       |
| Added Behats                      |
| Changelog updated                 |
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |
| Migration script                  |
| Tech Doc                          |

### CODE REVIEW

File: src/Oro/Bundle/DataGridBundle/Resources/config/services.yml [hunk 1]
```diff
@@ -53,7 +53,7 @@ services:
     oro_datagrid.twig.metadata:
         class: %oro_datagrid.twig.metadata.class%
         arguments:
-            - @oro_datagrid.datagrid.manager
+            - @service_container
```
Comment (jmleroux) [inline src/Oro/Bundle/DataGridBundle/Resources/config/services.yml]: To get rid of cirecular reference problem

**Comment by jmleroux** [comment]
This PR pass CE CI.
I will try a perfbench to see if it has some impacts

**Comment by jmleroux** [comment]
Using lazy service should avoid instantiation of LOTS of unused services at runtime.
Let's give it a try.

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php [hunk 2]
```diff
@@ -85,12 +83,36 @@ public function getGridData(\Twig_Environment $twig, $name, $params = [])
...
         ];
 
-        return $this->router->generate(self::ROUTE, $params);
+        return $this->getRouter()->generate(self::ROUTE, $params);
+    }
+
+    /**
+     * @return Manager
+     */
+    final protected function getDatagridManager()
```
Comment (nidup) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: What you try to ensure here, forbid to override this method in a subclass?

File: src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php [hunk 3]
```diff
@@ -133,26 +107,26 @@ public function deleteAction($id)
...
+     * @return Response
      */
-    public function getCheckidAction()
+    public function getCheckidAction(Request $request)
     {
         $entity = $this
             ->getDoctrine()
             ->getRepository('OroNavigationBundle:PageState')
-            ->findOneByPageHash(PageState::generateHash($this->getRequest()->get('pageId')));
+            ->findOneBy([
```
Comment (nidup) [inline src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php]: why we change this?

File: src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php [hunk 4]
```diff
@@ -162,17 +136,20 @@ protected function getManager()
...
      * Get entity by id
      *
+     * @param string|int $id
+     *
      * @return PageState
      */
     protected function getEntity($id)
     {
-        return $this->getDoctrine()->getRepository('OroNavigationBundle:PageState')->findOneById((int) $id);
+        return $this->getDoctrine()->getRepository('OroNavigationBundle:PageState')->findOneBy(['id' => (int)$id]);
```
Comment (nidup) [inline src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php]: same, why this change?

File: src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml [hunk 5]
```diff
@@ -13,7 +13,8 @@ parameters:
 services:
     # Dispatchers
     pim_datagrid.extension.mass_action.dispatcher:
-        class: %pim_datagrid.extension.mass_action.dispatcher.class%
+        class: '%pim_datagrid.extension.mass_action.dispatcher.class%'
+        lazy: true
```
Comment (nidup) [inline src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml]: still required with the service locator use?

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php (hunk 2 above)
Comment (jmleroux) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: yes

File: src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php (hunk 3 above)
Comment (jmleroux) [inline src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php]: getRequest is deprecated.
prepare the future. #boyScout

File: src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php (hunk 4 above)
Comment (jmleroux) [inline src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php]: magic calls are quite dégeu no ?

File: src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml (hunk 5 above)
Comment (jmleroux) [inline src/Pim/Bundle/DataGridBundle/Resources/config/mass_actions.yml]: absolutly not.
not even used.
good catch

File: src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php (hunk 4 above)
Comment (nidup) [inline src/Oro/Bundle/NavigationBundle/Controller/Api/PagestateController.php]: Ooohh yes they are, I read too quickly, no pb to replace, you're right :)

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php (hunk 2 above)
Comment (jjanvier) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: why not private?

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php [hunk 6]
```diff
@@ -4,26 +4,24 @@
...
 
-class MetadataExtension extends \Twig_Extension
+class MetadataExtension extends Twig_Extension
 {
     const ROUTE = 'oro_datagrid_index';
 
-    /** @var Manager */
-    protected $manager;
+    /** @var ContainerInterface */
+    protected $serviceLocator;
```
Comment (jjanvier) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: you should call it $container imho

the fact it's implemented as by the service locator pattern in Symfony is a detail

File: src/Oro/Bundle/NavigationBundle/Event/RequestTitleListener.php [hunk 7]
```diff
@@ -3,24 +3,25 @@
...
 use Symfony\Component\HttpKernel\HttpKernel;
 
 class RequestTitleListener
 {
     /**
-     * @var TitleServiceInterface
+     * @var ContainerInterface
      */
-    private $service;
+    private $serviceLocator;
```
Comment (jjanvier) [inline src/Oro/Bundle/NavigationBundle/Event/RequestTitleListener.php]: same comment as before

File: src/Oro/Bundle/NavigationBundle/Event/ResponseHashnavListener.php [hunk 8]
```diff
@@ -11,19 +11,13 @@ class ResponseHashnavListener
     const HASH_NAVIGATION_HEADER = 'x-oro-hash-navigation';
 
     /**
-     * @var TokenStorageInterface
+     * @var ContainerInterface
      */
-    protected $tokenStorage;
+    protected $serviceLocator;
```
Comment (jjanvier) [inline src/Oro/Bundle/NavigationBundle/Event/ResponseHashnavListener.php]: same comment as before

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php (hunk 2 above)
Comment (juliensnz) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: This is a more general discussion and introduce it like that is not a good idea. We don't need it here so it should not be final (even if I really would like to see some final in our code)

File: src/Oro/Bundle/NavigationBundle/Event/ResponseHashnavListener.php [hunk 9]
```diff
@@ -11,19 +11,13 @@ class ResponseHashnavListener
...
-     * @var EngineInterface
-     */
-    protected $templating;
-
-    public function __construct(TokenStorageInterface $tokenStorage, EngineInterface $templating)
+    public function __construct(ContainerInterface $container)
     {
-        $this->tokenStorage = $tokenStorage;
-        $this->templating = $templating;
+        $this->serviceLocator = $container;
```
Comment (juliensnz) [inline src/Oro/Bundle/NavigationBundle/Event/ResponseHashnavListener.php]: ?

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php (hunk 6 above)
Comment (BitOne) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: Not exactly ;) In fact, when you use the container like that, you effectively use a service locator.
container is really tainted DI, and here we are not using DI, but Service locator.

**Comment by jjanvier** [comment]
I'd rename the $serviceLocator vars to $container.
Not sure about the final methods. Why don't put them private instead? And even maybe protected if the team doesn't want it. It's no big. The important is that it works well out of the box.

Could you please add some insights about the gain?

Very cool PR :+1: Thanks JM !

File: src/Oro/Bundle/NavigationBundle/Event/ResponseHistoryListener.php [hunk 10]
```diff
@@ -6,51 +6,40 @@
...
 use Oro\Bundle\NavigationBundle\Provider\TitleServiceInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
 use Symfony\Component\HttpKernel\HttpKernel;
 use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\User\User;
 
 class ResponseHistoryListener
```
Comment (juliensnz) [inline src/Oro/Bundle/NavigationBundle/Event/ResponseHistoryListener.php]: should be changed to a subscriber. It will avoid to user the container

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php (hunk 6 above)
Comment (jmleroux) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: Service locator or DiC difference is not on implementation, but in the way you use it.

Use it to inject your services and it's IoC
Inject it directly and it's a SL.

Hence the naming.

**Comment by jmleroux** [comment]
@jjanvier ,

The service getter cannot be private, because they should be accessible for children classes.
So they must be private to not be overriden.

Moreover, it's not pushed, but i've made the `$serviceLocator` private to totally encapsulate the service location and force integrator to use the final getters.

File: src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php (hunk 2 above)
Comment (BitOne) [inline src/Oro/Bundle/DataGridBundle/Twig/MetadataExtension.php]: The main idea is to avoid publishing the serviceLocator/container, so it makes sense to forbid overriding.
I'm not really for a general discussion on private/final stuff, as it doesn't really make sense. It's not a like a checkstyle or PMD rule than can be applied everywhere.
private/final can be used and discussed depending on the code produced.

Having a general discussion would end with a "it depends", as it's cannot be an absolute rule to use it everywhere or nowhere, as this will not make any sense.

**Comment by jmleroux** [comment]
A word on my naming choice:

```
    /** @var ContainerInterface */
    private $serviceLocator;

    public function __construct(ContainerInterface $container)
    {
        $this->serviceLocator = $container;
    }
```

so when i do `$this->serviceLocator = $container;` , i mean it.
I can even translate it in plain text : 

> inject the service container ad use it as a service locator.

thoughts ?

**Comment by nidup** [comment]
I would :heart: move on this topic, this PR is really nice and it will help us to fix several circular reference around ConnectorRegistry which occurs due to the use of Nelmio in one of my PR and one PR of @grena :wink:

If you've a moment this week, let's discuss IRL, fix few pbs and merge ;)

**Comment by jmleroux** [comment]
With pleasure @nidup , don't hesitate to ping me ;)

File: src/Oro/Bundle/NavigationBundle/Event/ResponseHashnavListener.php (hunk 9 above)
Comment (a2xchip) [inline src/Oro/Bundle/NavigationBundle/Event/ResponseHashnavListener.php]: Reasonable question

Comment (jmleroux) [inline src/Oro/Bundle/NavigationBundle/Event/ResponseHashnavListener.php]: I explained my choice but if you are more to think otherwise, i can change the naming without any problem. ;)

**Comment by nidup** [comment]
I fully agree to change these kernel listeners to inject container inside.

I understand that you named the variable $serviceLocator because you only to fetch services from and not register new one, however it may be confusing for teammates, so I'd rename to $container. I do agree with the use of "final + protected" to avoid override of methods but allow using these in subclasses and I do agree too on the "private" visibility of the container to avoid accessing it in subclasses.

Even if not directly related to the listener change, I do understand and I'm agree to change the Datagrid/Manager and the few Twig/Extensions, due to the datagrid compiler pass, we always load a bunch of services on every call.

Why do we remove the cache warmup? We should never need this kind of warm so it's interesting but I don't understand how it's related with the main concern of this PR :wink: 

GTM once $serviceLocalor renamed to $container and once I have an explaination about family warmup drop.

**Comment by jmleroux** [comment]
@nidup , the family warmup removal is indeed a side effect of my journey with blackfire.
It can remove it from the PR.

**Comment by nidup** [comment]
Ok, please extract the warmup to another PR runs a build on EE, then GTM, thx again, great PR :)
